src/ipips/ipip-0378.md

```diff
@@ -0,0 +1,111 @@
+
+## Motivation
+
+The IPFS interaction with DHT includes both read and write operations.
+A user can provide records, advertising the presence of content, as well as looking up providers for a given CID.
+
+The specification proposed by
+[IPIP-337](https://specs.ipfs.tech/ipips/ipip-0337) offers an idiomatic
+first-class support for offloading the lookup portion of this interaction onto
+other processes and/or servers.
+
+Following the same motivations that inspired IPIP-337, this document expands the HTTP APIs to also
+offload the ability to provide records noto a third-party system.
+
+## Detailed design
+
+HTTP POST support is added
+to the [Delegated Content Routing HTTP API](https://specs.ipfs.tech/routing/http-routing-v1/)
+at `/routing/v1/providers` and `/routing/v1/peers`, along with complimentary
+sections that outline known formats followed by example payload.
+
+## Design rationale
+
+The rationale for the design of `POST` operations closely follows the reasoning
+listed in [IPIP-337](https://specs.ipfs.tech/ipips/ipip-0337/).
+The design uses a human-readable request/response structure with extensibility in mind.
+The specification imposes no restrictions on the schema nor the protocol advertised in provider records.
+The hope is that such extensibility will encourage and inspire innovation for better transfer protocols.
+
+In order to reduce barrier for adoption, standard HTTP POST semantics are used,
+along with transport-agnostic signature of DAG-CBOR payload representation
+similar to the data field in IPNS. This ensures this IPIP does not introduce
+any new signature dependency because IPNS records are already supported on
+`/routing/v1` since [IPIP-379](https://specs.ipfs.tech/ipips/ipip-0379/).
+
+### User benefit
+
+Expanding the user benefits listed as part of
+[IPIP-337](https://specs.ipfs.tech/ipips/ipip-0337/#user-benefit), in the
+context of content routing write operations are typically more expensive than
+read operations. They involve book keeping such as TTL, gossip propagation,
+etc. Therefore, it is highly desirable to reduce the burden of advertising
+provider records onto the network by means of delegation through simple to use
+HTTP APIs.
+
+### Compatibility
+
+#### Backwards Compatibility
+
+##### DHT
+
+Records published through HTTP delegated routing will not benefit from implicit
+PeerID identity and libp2p handshake, and must be explicitly signed.
+
+At the time of writing of this IPIP the Amino DHT does not have a concept of
+explicit signature. That is why the `POST` APIs proposed here introduce a new
+DAG-CBOR data format for specifying provider records, and mechanism for signing
+them.
+
+##### Reframe
+
+The Reframe API was superseded by `/routing/v1`.
+See [IPIP-337/Backwards Compatibility](https://specs.ipfs.tech/ipips/ipip-0337/#backwards-compatibility).
+
+#### Forwards Compatibility
```

Comment:
TODO

As this IPIP is right now, the routing announcement signature is used only for the remote HTTP server to decide if announcement should be accepted or not. This blocks announcing with PeerID that you don't own.

However, when user asks for providers, they won't get original payload+signature back, and won't be able to verify provider records themselves. 

**We don't have to solve or require it in this IPIP,** but we should add section to "Forwards Compatibility" explaining if/how signature type introduced in this IPIP could be stored and returned to end clients when someoneone wants that functionality.

If we don't write it down, we will be having multiple competing conventions, because people who want end-to-end routing integrity will invent something.

---- 

Initial thought on this is that we already have a concept of opaque  metadata fields for opaque protocols in [Peer Schema](https://specs.ipfs.tech/routing/http-routing-v1/#peer-schema). 

If we reuse it here, this section could hint at doing something like:

```json
{
  "Schema": "peer",
  "ID": "bafz...",
  "Addrs": ["/ip4/..."],
  "Protocols": ["transport-bitswap", "signed-routing-v1"],
  "signed-routing-v1": {
    "Payload": "[mbase64-dag-cbor-blob]", 
    "Signature": "[mbase64-blob]"
  }
}
```

Any concerns  @masih @willscott @hacdias ? 
This is just for "Forwards Compatibility" section, but if we have it, we could also implement it in [someguy](https://github.com/ipfs-shipyard/someguy), allowing people end-to-end integrity when they self-host own router (e.g. in private swarms).

Reply:
No immediate concern on my part 👍